### ISSUE TITLE
RedirectUrl shown on expired Invoice

### ISSUE THREAD

**Issue by petzsch** [btcpayserver/btcpayserver#165]
According to the bitpay documentation the RedirectUrl should only be used for paid invoices (not expired ones)

> URL to redirect your shopper back to your website after a successful purchase. Be sure to include "http://" or "https://" in the url.

Ideally we would have a second redirectUrl for timed out invoices, but that would need to be optional as it would actually be an improvement over the BitPay API. :-)

Here a screenshot of my problem:
![image](https://user-images.githubusercontent.com/1374810/39893534-580964e8-54a4-11e8-8761-d4173c84b97d.png)

**Comment by petzsch** [comment]
I would propose to instead redirect the user to the Shop URL if an invoice is expired. That is if we don't implement a RedirectOnFailureUrl :-)

**Comment by astupidmoose** [comment]
Can confirm, failed payments link customer back to success page. 

I think having it link back to Shop URL on a fail is a cleaner method.

**Comment by hubiktomas** [comment]
On one payment gateway I have seen just single return URL and the return status was identified using parameter in the URL (expired/canceled/paid/...). Then it was up to the client side to take appropriate action based on this variable. (https://bitcoinpaymerchantapi.docs.apiary.io and older version https://bitcoinpaycom.docs.apiary.io)

**Comment by petzsch** [comment]
Any thoughts on this? @NicolasDorier

**Comment by NicolasDorier** [comment]
On changing for sending to website? Yeah seems good, if merchants want to go that way it is good for me.

@hubiktomas is also possibility but it put more burden on the plugin developer.

**Comment by petzsch** [comment]
Yeah I think it's an improvement over the current situation. Better would be the proposed RedirectOnFailureUrl. But that would mean changes to the BitPay php api toolkit (and others) and I guess we want to stay api compatible.

**Comment by NicolasDorier** [comment]
WI would rather implement @hubiktomas proposal than `RedirectOnFailureUrl`. Because in both case you can get the same feature, but @hubiktomas does not break the API.

**Comment by hubiktomas** [comment]
@NicolasDorier It depends. It will be more compatible with the current version as developers can just use come conditional statement and change text based on this variable. With different URL, they will have to play with routing a little bit. But not a substantial problem in both cases.

**Comment by NicolasDorier** [comment]
Well you can just remove the parameter before checking the checksum. :p

**Comment by petzsch** [comment]
It's a bit tricky with the invoice beeing a public interface anyway. Anyone can query the success URL for the API and set a payment to pending. (success will only happen after IPN completed notification).

The checksum checking is sadly enough out of my scope: [thx ResellerClub](http://manage.resellerclub.com/kb/servlet/KBServlet/faq411.html)

The Invoice ID as parameter in the return URL would allow me to validate the payment status and then construct the final redirect URL.

Long story short: We need a {placeholder} for status and hopefully invoice-id to put in the RedirectUrl if the user wishes to use that feature. :-)

**Comment by NicolasDorier** [comment]
I don't understand. If you use a placeholder, then you are unable to calculate the checksum in advance.

**Comment by petzsch** [comment]
For me there a 2 options:

* either sent the customer directly from BTCPay back to ResellerClub with status P or N + the correct checksum and all (very tricky and maybe open to manipulation as the redirectUrl can be querried over the public invoices/id API endpoint)
* sent the customer to an intermediate page with the invoice ID as parameter, then I can query the invoice status, match the invoice-ID with the transactionID from ResellerClub from my database and send the customer to the failure or pending page with correct checksum to ResellerClub.

I think I'm in favour of option 2.

The Payment/Page flow is a bit complex:

ResellerClub -> my-intermediate-page/createInvoice.php?lot-of-paramets-from-RC -> BTCPay -> my-intermediate-page/redirect.php?id=invoice-id -> ResellerClub/?status=P&checksum and so forth.

Sorry I'm getting offtopic here. If it's not to much effort to also implement the invoice-id as a placeholder in the redirect-url then that would be awesome. Else I'll have to try and put that information into SESSION vars in my intermediate-page and have it check for those when the visitor comes back.

**Comment by NicolasDorier** [comment]
Ok that seems cool. Can you open a separate issue for having those placeholders?

**Comment by petzsch** [comment]
I now rely on the redirectUrl beeing shown on failed invoices. If we do change that behaviour my implementation will break. :-(

**Comment by NicolasDorier** [comment]
@petzsch are you using the Website URL? what I can do is to use redirectUrl if no website url is set.

**Comment by petzsch** [comment]
I have it configured, but can remove it ... it's not really used.

**Comment by NicolasDorier** [comment]
mmh on another hand I don't like it... this would not be documented and hackish...

**Comment by petzsch** [comment]
I'm not so sure which URL bitpay uses for the redirect on timeout. To me it looks like they could also be using the redirectURL, see this [test invoice](https://bitpay.com/invoice?id=j5CU9ee21noWonWR63EEd) as an example: it redirects to a success page even though it's timed out.

I'm unsure what to make of all of this: (a) We could leave it as is and be 100% bitpay compatible, (b) we could add placeholders in the current situation so that the target page can be customizable, (c) we could change the behaviour to how it was documented in the API doc (website url on failure, else redirectURL). Leaving out option d here because adding a ´RedirectOnFailureUrl´ would break the api.

Perhaps we should start a vote here?

I'm currently tending toward option b. :-)

**Comment by NicolasDorier** [comment]
I quite like (b) as well. It seems to me to be the least invasive change.

**Comment by hubiktomas** [comment]
I like the (b) option as well. :) It is quite flexible and not so invasive. It would be also great if the developer could add custom vars to the return URL via API. For instance order_id, some security keys, etc. And invoice status change notify URL as well.

**Comment by NicolasDorier** [comment]
> It would be also great if the developer could add custom vars to the return URL via AP

He already can.

**Comment by hubiktomas** [comment]
@NicolasDorier - Ah, ok, thank you, sorry. I did not have time to dig deeper in the API yet.

So for me placeholder with return status is enough to allow simple solutions work and will not break older implementations.

**Comment by NicolasDorier** [comment]
Closing this as https://github.com/btcpayserver/btcpayserver/issues/169 fixed it
